# HG changeset patch
# User Ping Chen <remotenonsense@gmail.com>
# Date 1665496501 -7200
# Node ID dfc327877efcb0b1ab4588b71298e2162168d22a
# Parent  2869d7d4ff9ad3390a4cc15dfbdc52d57802c643
Bug 1794185 - Fix "Get Next n Messages" for nntp-js. r=mkmelin

Update nsIDBFolderInfo.knownArtsSet based on XOVER range, doesn't matter if artciles in the range don't exist on the server.

Differential Revision: https://phabricator.services.mozilla.com/D158898

diff --git a/mailnews/news/src/NntpClient.jsm b/mailnews/news/src/NntpClient.jsm
--- a/mailnews/news/src/NntpClient.jsm
+++ b/mailnews/news/src/NntpClient.jsm
@@ -548,16 +548,17 @@ class NntpClient {
 
   /**
    * A transient action to consume the status line of XOVER response.
    * @param {NntpResponse} res - XOVER response received from the server.
    */
   _actionXOverResponse(res) {
     if (res.status == 224) {
       this._nextAction = this._actionReadXOver;
+      this._newsGroup.addKnownArticles(this._startArticle, this._endArticle);
       this._actionReadXOver(res);
     } else {
       // Somehow XOVER is not supported by the server, fallback to use HEAD to
       // fetch one by one.
       this._actionHead();
     }
   }
 
diff --git a/mailnews/news/src/NntpNewsGroup.jsm b/mailnews/news/src/NntpNewsGroup.jsm
--- a/mailnews/news/src/NntpNewsGroup.jsm
+++ b/mailnews/news/src/NntpNewsGroup.jsm
@@ -148,17 +148,23 @@ class NntpNewsGroup {
     msgHdr.subject = subject;
     msgHdr.author = from;
     msgHdr.date = new Date(date).valueOf() * 1000;
     msgHdr.messageId = messageId;
     msgHdr.setReferences(references);
     msgHdr.messageSize = bytes;
     msgHdr.lineCount = lines;
     this._msgHdrs.push(msgHdr);
-    this._knownKeySet.add(articleNumber);
+  }
+
+  /**
+   * Add a range (usually XOVER range) to the known key set.
+   */
+  addKnownArticles(start, end) {
+    this._knownKeySet.addRange(start, end);
   }
 
   /**
    * Finish processing XOVER responses.
    */
   finishProcessingXOver() {
     this._runFilters();
     let groupInfo = this._db.dBFolderInfo;
